### PR TITLE
storage: Expose Stratis virtual filesystem sizes

### PR DESCRIPTION
**Pull request by mvollmer**
Instead of the "Managed filesystem sizes" concept. This makes Cockpit
easier to discover when you know the real Stratis concepts and exposes
its full capabilities.

Demo: https://youtu.be/E1zqhAowus8 
Longer exploration: https://youtu.be/1W6QPNPwT7Q

### CODE REVIEW

File: pkg/storaged/client.js [hunk 1]
```diff
@@ -1399,17 +1399,13 @@ function stratis3_start() {
...
-                    // "pool.Overprovisioning = flag" to just work,
-                    // but we want to catch any error ourselves, and
-                    // we want to wait for the method call to
-                    // complete.
-                    return stratis.call(pool.path, "org.freedesktop.DBus.Properties", "Set",
-                                        ["org.storage.stratis3.pool." + stratis3_interface_revision,
-                                            "Overprovisioning",
-                                            cockpit.variant("b", flag)
-                                        ]);
+                client.stratis_set_property = (proxy, prop, sig, value) => {
```
Comment (cockpituous) [inline pkg/storaged/client.js:1402]: This added line is not executed by any test.

File: pkg/storaged/client.js [hunk 2]
```diff
@@ -1399,17 +1399,13 @@
...
+                    // DBusProxy is smart enough to allow "proxy.Prop
+                    // = value" to just work, but we want to catch any
+                    // error ourselves, and we want to wait for the
+                    // method call to complete.
+                    return stratis.call(proxy.path, "org.freedesktop.DBus.Properties", "Set",
+                                        [proxy.iface, prop, cockpit.variant(sig, value)]);
```
Comment (cockpituous) [inline pkg/storaged/client.js:1408]: These 2 added lines are not executed by any test.

File: pkg/storaged/stratis/create-dialog.jsx [hunk 3]
```diff
@@ -130,11 +130,13 @@ export function create_stratis_pool() {
                                                                  clevis_info ? [true, clevis_info] : [false, ["", ""]])
                                 .then(std_reply)
                                 .then(result => {
-                                    if (vals.managed && vals.managed.on && result[0]) {
+                                    if (vals.overprov && !vals.overprov.on && result[0]) {
```
Comment (cockpituous) [inline pkg/storaged/stratis/create-dialog.jsx:133]: This added line is not executed by any test.

File: pkg/storaged/stratis/create-dialog.jsx [hunk 4]
```diff
@@ -130,11 +130,13 @@
...
                                         const path = result[1][0];
                                         return client.wait_for(() => client.stratis_pools[path])
                                                 .then(pool => {
-                                                    return client.stratis_set_overprovisioning(pool, false);
+                                                    return client.stratis_set_property(pool,
+                                                                                       "Overprovisioning",
+                                                                                       "b", false);
```
Comment (cockpituous) [inline pkg/storaged/stratis/create-dialog.jsx:139]: These 3 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 5]
```diff
@@ -206,6 +194,44 @@ const StratisFilesystemCard = ({
         });
     }
 
+    function set_limit() {
+        dialog_open({
+            Title: _("Set limit of virtual filesystem size"),
+            Fields: [
+                CheckBoxes("size_options", _("Options"),
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:201]: These 5 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 6]
```diff
@@ -206,6 +194,44 @@
...
+                           {
+                               value: {
+                                   "custom_limit": fsys.SizeLimit[0],
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:204]: These 2 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 7]
```diff
@@ -206,6 +194,44 @@
...
+                               },
+                               fields: [
+                                   { tag: "custom_limit", title: _("Limit virtual filesystem size") },
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:207]: These 2 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 8]
```diff
@@ -206,6 +194,44 @@
...
+                               ]
+                           }),
+                SizeSlider("limit", _("Virtual size limit"),
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:210]: This added line is not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 9]
```diff
@@ -206,6 +194,44 @@
...
+                           {
+                               visible: vals => vals.size_options.custom_limit,
+                               value: fsys.SizeLimit[0] && Number(fsys.SizeLimit[1]),
+                               min: Number(fsys.Size),
+                               max: pool.Overprovisioning ? stats.pool_total : stats.pool_free + Number(fsys.Size),
+                               allow_infinite: true,
+                               round: 512
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:217]: These 6 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 10]
```diff
@@ -206,6 +194,44 @@
...
+                           }),
+            ],
+            Action: {
+                Title: _("Set"),
+                action: async function (vals) {
+                    await client.stratis_set_property(fsys,
+                                                      "SizeLimit",
+                                                      "(bs)",(vals.size_options.custom_limit
+                                                              ? [true, vals.limit.toString()]
+                                                              : [false, ""]));
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:227]: These 8 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 11]
```diff
@@ -224,13 +250,16 @@
...
+                        <StorageUsageBar stats={[Number(fsys.Used[0] && Number(fsys.Used[1])), stats.pool_total]}
                                              critical={1} total={stats.fsys_total_used} offset={offset} />
-                            : <StorageUsageBar stats={[Number(fsys.Used[0] && Number(fsys.Used[1])), Number(fsys.Size)]}
-                                             critical={0.95} />)
-                        }
                     </StorageDescription>
+                    <StorageDescription title={_("Virtual size")}
+                                        value={fmt_size(Number(fsys.Size))}/>
+                    <StorageDescription title={_("Virtual size limit")}
+                                        value={fsys.SizeLimit[0] ? fmt_size(Number(fsys.SizeLimit[1])) : _("none")}
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:259]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 12]
```diff
@@ -93,11 +92,31 @@ function create_fs(pool) {
...
+                               { tag: "custom_size", title: _("Specify initial virtual filesystem size") },
+                               { tag: "custom_limit", title: _("Limit virtual filesystem size") },
+                           ]
+                       }),
+            SizeSlider("size", _("Initial virtual size"),
+                       {
+                           visible: vals => vals.size_options.custom_size,
                            min: fsys_min_size,
-                           max: stats.pool_free,
+                           max: pool.Overprovisioning ? stats.pool_total : stats.pool_free,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:110]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 13]
```diff
@@ -93,11 +92,31 @@
...
+                           allow_infinite: pool.Overprovisioning,
+                           round: 512
+                       }),
+            SizeSlider("limit", _("Virtual size limit"),
+                       {
+                           visible: vals => vals.size_options.custom_limit,
+                           min: fsys_min_size,
+                           max: pool.Overprovisioning ? stats.pool_total : stats.pool_free,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:118]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 14]
```diff
@@ -115,15 +134,15 @@
...
-                        .then(result => {
-                            if (result[0])
-                                return set_mount_options(result[1][0][0], vals, forced_options);
-                            else
-                                return Promise.resolve();
-                        });
+            action: async function (vals) {
+                let size_spec = [false, ""], limit_spec = [false, ""];
+                if (vals.size_options.custom_size)
+                    size_spec = [true, vals.size.toString()];
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:140]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 15]
```diff
@@ -115,15 +134,15 @@
...
+                if (vals.size_options.custom_limit)
+                    limit_spec = [true, vals.limit.toString()];
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:142]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 16]
```diff
@@ -514,11 +527,18 @@
...
+                      <StorageDescription title={_("Usage")}>
+                          <StorageUsageBar stats={use} critical={0.95} />
+                      </StorageDescription>
                     }
+                    <StorageDescription title={_("Overprovisioning")}>
+                        <StorageOnOff state={pool.Overprovisioning}
+                                      aria-label={_("Allow overprovisioning")}
+                                      onChange={() => client.stratis_set_property(pool,
+                                                                                  "Overprovisioning",
+                                                                                  "b", !pool.Overprovisioning)} />
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:540]: These 3 added lines are not executed by any test.

File: pkg/storaged/stratis/filesystem.jsx [hunk 17]
```diff
@@ -206,6 +199,42 @@ const StratisFilesystemCard = ({
...
+                               fields: [
+                                   { tag: "custom_limit", title: _("Limit virtual filesystem size") },
+                               ]
+                           }),
+                SizeSlider("limit", _("Virtual size limit"),
+                           {
+                               visible: vals => vals.size_options.custom_limit,
+                               value: fsys.SizeLimit[0] && Number(fsys.SizeLimit[1]),
+                               min: Number(fsys.Size),
+                               max: pool.Overprovisioning ? stats.pool_total : stats.pool_free + Number(fsys.Size),
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:220]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 18]
```diff
@@ -86,18 +85,40 @@ function create_fs(pool) {
...
+                           fields: [
+                               { tag: "custom_size", title: _("Specify initial virtual filesystem size") },
+                               { tag: "custom_limit", title: _("Limit virtual filesystem size") },
+                           ]
+                       }),
+            SizeSlider("size", _("Initial virtual size"),
+                       {
+                           visible: vals => vals.size_options.custom_size,
+                           min: fsys_min_size,
+                           max: pool.Overprovisioning ? stats.pool_total : stats.pool_free,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:112]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 19]
```diff
@@ -86,18 +85,40 @@
...
+                           allow_infinite: pool.Overprovisioning,
+                           round: 512
+                       }),
+            SizeSlider("limit", _("Virtual size limit"),
                        {
-                           visible: () => managed_fsys_sizes,
+                           visible: vals => vals.size_options.custom_limit,
                            min: fsys_min_size,
-                           max: stats.pool_free,
+                           max: pool.Overprovisioning ? stats.pool_total : stats.pool_free,
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:120]: This added line is not executed by any test.

File: pkg/storaged/stratis/pool.jsx [hunk 20]
```diff
@@ -514,10 +532,25 @@
...
+                      <StorageDescription title={_("Usage")}>
+                          <StorageUsageBar stats={use} critical={0.80} />
+                      </StorageDescription>
+                    }
+                    <StorageDescription title={_("Overprovisioning")}>
+                        <StorageOnOff state={pool.Overprovisioning}
+                                      aria-label={_("Allow overprovisioning")}
+                                      onChange={() => client.stratis_set_property(pool,
+                                                                                  "Overprovisioning",
+                                                                                  "b", !pool.Overprovisioning)}
```
Comment (cockpituous) [inline pkg/storaged/stratis/pool.jsx:545]: These 3 added lines are not executed by any test.

**Comment by mvollmer** [comment]
Let's get this in. The UI here assumes that people are experts in Stratis and know exactly what they are doing and don't need any help or explanations from Cockpit.

**Comment by mvollmer** [comment]
> No strong argument against or for.

Yeah, it's meh all around.

> There is one tiny `codeql` warning however which seems nice to fix.

There's lots of linting to be done, but I think the `Variable m is not used. ` warning has been fixed.

**Comment by mvollmer** [comment]
> but I think the `Variable m is not used. ` warning has been fixed.

Ohh: there were two "unused `m`" warnings, and I only fixed one, but the comments here show the wrong one as fixed! I am always suprised at how fragile these shiny web things actually are.

File: pkg/storaged/stratis/filesystem.jsx [hunk 21]
```diff
@@ -206,6 +200,42 @@ const StratisFilesystemCard = ({
...
```
Comment (cockpituous) [inline pkg/storaged/stratis/filesystem.jsx:221]: This added line is not executed by any test.

**Review by jelly — APPROVED** [review]

**Pull request merged**